feelpp/feel/feelcrb/pbdw.hpp

```diff
@@ -0,0 +1,355 @@
+    return I;
+}
+
+template<typename RBSpace>
+void
+PBDW<RBSpace>::saveDB()
+{
+    if ( this->worldComm().isMasterRank() )
+    {
+        if( fs::exists(this->dbLocalPath()) )
+        {
+            if( !fs::is_directory(this->dbLocalPath()) )
+                throw std::exception();
+        }
+        else
+            fs::create_directories(this->dbLocalPath());
+
+        json j;
+        j["name"] = this->name();
+        std::time_t t = std::time(nullptr);
+        std::stringstream ss;
+        ss << std::put_time(std::localtime(&t), "%c %Z");
+        j["date"] = ss.str();
+        j["mesh"] = this->absoluteMeshFilename();
+        std::ofstream o(this->absoluteJsonFilename());
+        o << j << std::endl;
+
+        fs::ofstream ofs( this->absoluteDbFilename() );
+        if ( ofs )
+        {
+            Feel::cout << "saving DB at " << this->absoluteDbFilename() << std::endl;
+            boost::archive::binary_oarchive oa( ofs );
+            oa << this->dimensionN();
+            oa << this->dimensionM();
+            oa << M_matrix;
+        }
+    }
+    if( ! fs::exists(this->absoluteMeshFilename()) )
+    {
+        Feel::cout << "Saving mesh to " << this->absoluteMeshFilename() << std::endl;
+        M_XR->functionSpace()->mesh()->saveHDF5( this->absoluteMeshFilename() );
+    }
+    fs::ofstream ofsp( this->absoluteDbFilenameProc() );
+    if( ofsp )
+    {
+        boost::archive::binary_oarchive oa( ofsp );
+        for( int n = 0; n < M_N; ++n )
+            oa << M_XR->primalBasisElement(n);
+        for( int m = 0; m < M_M; ++m )
+        {
+            auto t = M_sigmas[m]->type();
+            oa << t;
+            oa << M_sigmas[m]->name();
+            oa << M_sigmas[m]->position();
+            if( t == "pointwise" )
+            {
+            }
+            else if( t == "gaussian" )
+            {
+                auto s = std::dynamic_pointer_cast<SensorGaussian<space_type>>(M_sigmas[m]);
+                oa << s->radius();
+            }
+        }
+    }
+}
+
+template<typename RBSpace>
+void
+PBDW<RBSpace>::loadDB( std::string const& filename, crb::load l )
+{
+    if( !fs::exists(filename) )
+        return;
+
+    fs::ifstream ifs( filename );
+    if ( ifs )
+    {
+        Feel::cout << "loading DB at " << filename << std::endl;
+        boost::archive::binary_iarchive ia( ifs );
+        if( l > crb::load::none )
+        {
+            ia >> M_N;
+            ia >> M_M;
+            M_matrix.resize(M_M+M_N, M_M+M_N);
+            ia >> M_matrix;
+        }
+    }
+    if( l > crb::load::rb )
+    {
+        if( ! M_XR )
+        {
+            std::ifstream i(this->absoluteJsonFilename());
+            json j = json::parse(i);
+            auto meshfilename = j["mesh"].get<std::string>();
+            auto mesh = loadMesh(_mesh=new mesh_type, _filename=meshfilename);
+            auto Xh = space_type::New(mesh);
+            M_XR = std::make_shared<reducedspace_type>(Xh);
```

Comment:
why should PBDW take care of the serializing the reduced space ?
how about letting the reduced space do it  itself ?

Reply:
the reduced space does not have a serialization implemented, idem for the sensors

Reply:
we should add that no ?

Reply:
in terms of design that is better and can be reused

Reply:
I agree but I think that we need to think about it, for example if we have a function space on range, how do we rebuild it ?
We need to save the mesh and the range, no ?